README
------------------------------------------------------------------------
												DO NOT DISTRIBUTE
------------------------------------------------------------------------

ApplePushNotification
=====================

This plugin helps you using the Apple Push Notification system.

Once you have the certificate from Apple for your application, export your key
and the apple certificate as p12 files and run :

$ openssl pkcs12 -in cert.p12 -out apple_push_notification.pem -nodes -clcerts

put apple_push_notification.pem in config/

Example
=======

$ rake apn:migrate

$ ./script/console
>> a = ApplePushNotification.new
>> a.device_token = "c079040a 7d78c3de a68c813e 6b0c1fc1 e5d03d38 b661f715 XXXXXXXX XXXXXXXX"
>> a.badge = 5
>> a.sound = true
>> a.alert = "foobar"
>> a.send_notification
=> nil

Copyright (c) 2009 Fabien Penso.

Will be released under the MIT license once Apple release its SDK.

------------------------------------------------------------------------
												DO NOT DISTRIBUTE
------------------------------------------------------------------------


CHANGELOG
0.1.1 (November 10th, 2009)

* restructure to make model more flexible

* moved migration to generator 
 
 
0.1.0 (November 10th, 2009)
 
* initial release


test/apple_push_notification_test.rb
require 'test_helper'

class ApplePushNotificationTest < ActiveSupport::TestCase
  # Someone could be awesome and write a test
  test "the truth" do
    assert true
  end
end


lib/apple_push_notification/base.rb
require 'socket'
require 'openssl'

module ApplePushNotification
  class Base
  	attr_accessor :paylod, :sound, :badge, :alert
  	attr_accessible :device_token
	
  	PORT = 2195
	
  	cattr_accessor :enviroment
  	self.enviroment = :development

  	validates_uniqueness_of :device_token

  	def send_notification
      raise "Missing cert: #{_path}" unless @@cert

  		ctx = OpenSSL::SSL::SSLContext.new
  		ctx.key = OpenSSL::PKey::RSA.new(@@cert)
  		ctx.cert = OpenSSL::X509::Certificate.new(@@cert)

  		s = TCPSocket.new(@@host, PORT)
  		ssl = OpenSSL::SSL::SSLSocket.new(s, ctx)
  		ssl.sync = true
  		ssl.connect

  		ssl.write(self.apn_message_for_sending)

  		ssl.close
  		s.close

  	rescue SocketError => error
  		raise "Error while sending notifications: #{error}"
  	end
	
  	def self.enviroment= enviroment
  	  @@enviroment = enviroment.to_sym
      @@host = self.production? ? "gateway.push.apple.com" : "gateway.sandbox.push.apple.com"
    	cert = self.production? ? "apn_production.pem" : "apn_development.pem"
    	path = File.join(File.expand_path(RAILS_ROOT), "config", "certs", cert)
      @@cert = File.exists?(path) ? File.read(path) : nil
    end
	
  	def self.development?
  	  @@enviroment != :production
    end
	
  	def self.production?
  	  @@enviroment == :production
    end

  	def self.send_notifications(notifications)	  
  		ctx = OpenSSL::SSL::SSLContext.new
  		ctx.key = OpenSSL::PKey::RSA.new(@@cert)
  		ctx.cert = OpenSSL::X509::Certificate.new(@@cert)

  		s = TCPSocket.new(@@host, PORT)
  		ssl = OpenSSL::SSL::SSLSocket.new(s, ctx)
  		ssl.sync = true
  		ssl.connect

  		for notif in notifications do
  			ssl.write(notif.apn_message_for_sending)
  		end

  		ssl.close
  		s.close
  	rescue SocketError => error
  		raise "Error while sending notifications: #{error}"
  	end

  	protected

  	def to_apple_json
  	  json = self.apple_array.to_json
  		logger.debug "Sending #{json}"
  		json
  	end

  	def apn_message_for_sending
  		json = self.to_apple_json
  		message = "\0\0 #{self.device_token_hexa}\0#{json.length.chr}#{json}"
  		raise "The maximum size allowed for a notification payload is 256 bytes." if message.size.to_i > 256
  		message
  	end

  	def device_token_hexa
  		[self.device_token.delete(' ')].pack('H*')
  	end

  	def apple_array
  		result = {}
  		result['aps'] = {}
  		result['aps']['alert'] = alert if alert
  		result['aps']['badge'] = badge.to_i if badge
  		result['aps']['sound'] = sound if sound and sound.is_a? String
  		result['aps']['sound'] = 'default' if sound and sound.is_a? TrueClass
  		result
  	end
  end
end

##
# Add acts_as_push method to ActiveRecord::Base so ApplePushNotification is accessible.
#
ActiveRecord::Base.class_eval do
  
  ##
  # Set attribute names and include the ApplePushNotification module.
  #
  def self.acts_as_push(enviroment = :development, options = {})
    class_inheritable_reader :apn_options
    write_inheritable_attribute :apn_options, {
      :enviroment   => enviroment,
      :certificate  => options[:certificate]
    }
    include ApplePushNotification
  end
end


lib/apple_push_notification.rb
require 'socket'
require 'openssl'

module ApplePushNotification

  def self.extended(base)
    # Added device_token attribute if not included by acts_as_pushable
    unless base.respond_to?(:acts_as_push_options)
      base.class_eval do
        attr_accessor :device_token
      end
    end
  end

  APN_PORT = 2195
  @@apn_cert = nil
  @@apn_host = nil
  
  def self.apn_enviroment
    @@apn_enviroment
  end
  
  def self.apn_development?
    @@apn_enviroment != :production
  end

  def self.apn_production?
    @@apn_enviroment == :production
  end
  
  def self.apn_enviroment= enviroment
    @@apn_enviroment = enviroment.to_sym
    @@apn_host = self.apn_production? ? "gateway.push.apple.com" : "gateway.sandbox.push.apple.com"
    cert = self.apn_production? ? "apn_production.pem" : "apn_development.pem"
    path = File.join(File.expand_path(RAILS_ROOT), "config", "certs", cert)
    @@apn_cert = File.exists?(path) ? File.read(path) : nil
    raise "Missing apple push notification certificate in #{path}" unless @@apn_cert
  end
  
  self.apn_enviroment = :development
  
  def send_notification options
    raise "Missing apple push notification certificate" unless @@apn_cert

    ctx = OpenSSL::SSL::SSLContext.new
    ctx.key = OpenSSL::PKey::RSA.new(@@apn_cert)
    ctx.cert = OpenSSL::X509::Certificate.new(@@apn_cert)

    s = TCPSocket.new(@@apn_host, APN_PORT)
    ssl = OpenSSL::SSL::SSLSocket.new(s, ctx)
    ssl.sync = true
    ssl.connect

    ssl.write(self.apn_message_for_sending(options))
    ssl.close
    s.close
  rescue SocketError => error
    raise "Error while sending notifications: #{error}"
  end
  
  def self.send_notification token, options = {}
    d = Object.new
    d.extend ApplePushNotification
    d.device_token = token
    d.send_notification options
  end

  protected

  def apn_message_for_sending options
    json = ApplePushNotification::apple_json_array options
    message = "\0\0 #{self.device_token_hexa}\0#{json.length.chr}#{json}"
    raise "The maximum size allowed for a notification payload is 256 bytes." if message.size.to_i > 256
    message
  end

  def device_token_hexa
    # Use `device_token` as the method to get the token from
    # unless it is overridde from acts_as_pushable
    apn_token_field = "device_token"
    if respond_to?(:acts_as_push_options)
      apn_token_field = acts_as_push_options[:device_token_field]
    end
    token = send(apn_token_field.to_sym)
    raise "Cannot send push notification without device token" if !token || token.empty?
    [token.delete(' ')].pack('H*')
  end

  def self.apple_json_array options
    result = {}
    result['aps'] = {}
    result['aps']['alert'] = options[:alert].to_s if options[:alert]
    result['aps']['badge'] = options[:badge].to_i if options[:badge]
    result['aps']['sound'] = options[:sound] if options[:sound] and options[:sound].is_a? String
    result['aps']['sound'] = 'default' if options[:sound] and options[:sound].is_a? TrueClass
    result.to_json
  end
    
end

require File.dirname(__FILE__) + "/apple_push_notification/acts_as_pushable"
